File: content/problems/0202-happy-number/solutions/floyd-cycle/Solution.java
class Solution {
    public boolean isHappy(int n) {
        int slow = n, fast = next(n);
        while (fast != 1 && slow != fast) { slow = next(slow); fast = next(next(fast)); }
        return fast == 1;
    }
    private int next(int value) { int sum = 0; while (value > 0) { int digit = value % 10; sum += digit * digit; value /= 10; } return sum; }
}

File: content/problems/0202-happy-number/solutions/floyd-cycle/Solution.kt
class Solution {
    fun isHappy(n: Int): Boolean {
        var slow = n; var fast = next(n)
        while (fast != 1 && slow != fast) { slow = next(slow); fast = next(next(fast)) }
        return fast == 1
    }
    private fun next(input: Int): Int { var value = input; var sum = 0; while (value > 0) { val digit = value % 10; sum += digit * digit; value /= 10 }; return sum }
}
